### ISSUE TITLE
Bug in the old 1983 game "Castle Wolfenstein"

### ISSUE THREAD

**Issue by Jkapp76** [86Box/86Box#4947]
### What happened?

This game starts up and shows the intro screen, then immediately says "you're caught!" and goes back to the intro screen.

Same with PCJr and Tandy 1000. (This game does make use of PCJr's 16 colors only in the title screen)

This game had the same issue with dosbox, but the dosbox staging project fixed it and the newer versions of Staging play it well.
The game recommends PCDos 2. I booted into this version and it still does the same. PCdos2 might have been the newest Dos version in 83.

The joystick does not work on this game either, it just moves in a diagonal direction and can't be adjusted.

The sequel, "Beyond Castle Wolfenstein" is a booter game and plays fine.

### Configuration file

```ini
PCJr and PCXT and Tandy 1000. 
I tried CGA with the the XT.
```

### Operating system

Windows 11

### CPU

intel i5

### 86Box version

newest stable release. (and a few others that work the same)

### Build architecture

Windows - x64 (64-bit)

### Build type

- [ ] New recompiler
- [ ] Debug build

### Download source

Official website (Jenkins, GitHub)

### Additional context

I collect very old PC games and booter disks if you ever need that type of testing. I like the really old stuff.

**Comment by Pellzi** [comment]
Have you tried out the newest nightly build (Approximative 6247)?
OBattler does a such good work for us all, that the version nummers counts very fast up.
Faster then i ever seen before for a program in my live.

To do so, if you don´t want to compile the newest 86Bos version by yourself:
Got to : https://86box.net/
Press "Download Experimental", there "Last Sucessful Artifacts, there Old Compiler (recommanded), there "Windows - x64(64-Bit)".
This downloads the newest Build with the newest Debug Strings from deep inside of the very complicated System of DOOS, and his Drivers for different Hardware.
Save this zip File somewhere on your System, unpack the zip File, copy the three Files from inside this Folder, you just produce, and paste them into your 86Box Folder, where the Originals are.
(Making a copy of them makes no sense, for me. Because, the Debug Level is much more less of these now replaced Files, that you have just updatet with this procdure written above.) 
Start 86Box, and try out, if this Faults occure also in the newest Nightly Build Version.
And please, if you done it, write a short feedback here.

By the Way to OBattler: Many Thanks for your Debug Work for the Project 86Box!
It's almost unbelievable that one Person is able to debug DOS problems.
This work you do now, should be done 25 or more Years before by Microsoft and Harware Manufactures, which haven´t done it until now.
I´m nearly absolute sure, that this problem with Castle Wolfenstein comes not from the Emalator 86Box itself.
Instead, it´s a very old problem, never solved before, for the real Hardware, and also not for a Real Emulator of it.
If you go on with your work, i see a possability, that you reincarnate DOS from his death because of all the different Bugs into it. Which cames from fault programming the Hardware during many Years in the Time before Millenium.

**Comment by OBattler** [comment]
@Jkapp76 Are you, by any chance, running an cracked copy of the game? I would suggest sourcing an 86F image of it and trying that.

**Comment by eddmanx** [comment]
And this: https://forum.vcfed.org/index.php?threads/cannot-get-castle-wolfenstein-working-in-ibm-pc.63608/

**Comment by Jkapp76** [comment]
My game is the original, not cracked.

The game will boot and work only with IBM PC and PC DOS 2.1. So this is good.
The game fails on PCJr using the same dos setup. PCJr is the better version that uses 16 colors.

But the game will now work with the specific setup I mentioned.

**Comment by OBattler** [comment]
On PCjr, it could be a RAM issue - we do not yet emulate the memory sidecar that would allow more than 128 kB of RAM.

**Comment by Jkapp76** [comment]
https://github.com/dosbox-staging/dosbox-staging/pull/3460

Here's a link to where dosbox staging worked on the PCJr's memory to get it compatible with pretty much every game made for it. This may or may not relate to 86Box, but it shows that there was something different about only the pcjr system that needed its own memory tweaks.

Also, I have never got the joystick to work with this game. I used an old analog stick on the real system, but the joystick does not work with 86box or dosbox.

**Comment by OBattler** [comment]
I'm going to have to look at the IBM PCjr Technical Reference for that, to see what I'm missing.

**Comment by RogerSanders** [comment]
This was caused by OS anti-tamper protection and/or debugger detection in the original Castle Wolfenstein game. I've provided technical details on the protection mechanisms and a patch for it here:
https://www.vogons.org/viewtopic.php?t=106714

**Comment by OBattler** [comment]
So basically, this can be closed as NOTABUG - anti-tamper protection and/or debugger detection that's prone to misfiring.

**Comment by eddmanx** [comment]
> So basically, this can be closed as NOTABUG - anti-tamper protection and/or debugger detection that's prone to misfiring.

Shouldn't it work on PCjr on DOS 2.1 though? Jkapp76 said the game works in 86box using DOS 2.1 using another setup.

Or perhaps the copy protection also expects a certain hardware?

**Comment by OBattler** [comment]
The thing is, with how much RAM does it work on a real PCjr, that is, assuming it does - so far, no evidence has been provided from real hardware.

**Comment by Jkapp76** [comment]
I can verify that the python patch mentioned above fixes the issue. The game is playable now. 

I still can't get a joystick to work properly. It pulls to the top left corner, but the game plays with keyboard.

The game will load from a bootable dos2.1 disk, but with the patch it will run from a hard drive and never fails.

DosBOX Staging added a feature to their PCJr video mode, the ability to adjust the brown/yellow with a simulated knob. I notice on 86Box the yellow soldiers appear brown in 86box and need this same adjustment. Also the bombs on the game "Jumpman" are brown and should be yellow. Staging fixed this color issue with the ability to change the brown\yellow.

**Comment by eddmanx** [comment]
Could you upload an image of the game?

**Comment by Jkapp76** [comment]
Do you mean a disk image? Or an image of the game running?

**Comment by Jkapp76** [comment]
![Image](https://github.com/user-attachments/assets/90c7194f-66af-41c2-ab64-46b83d9ae95a)

Here's a screenshot from DOSBox Staging showing the proper yellow soldier colors.

**Comment by OBattler** [comment]
Is that the unpatched game?

**Comment by OBattler** [comment]
Also, I'm quite sure the real PCjr monitor shows brown, not dark yellow, and the soldiers are quite assuredly supposed to be brown - they are most likely supposed to be SA (Sturmabteilung, Stormtroopers), who in real life, wore brown uniforms, hence the name brownshirts. Dark yellow makes no sense here.

**Comment by eddmanx** [comment]
> Do you mean a disk image? Or an image of the game running?

Disk. It would help with testing on dev builds and real hardware.

**Comment by OBattler** [comment]
Also, here's a photo of the real PCjr, showing brown:

![Image](https://github.com/user-attachments/assets/5c5946c6-c495-4ffa-bcc1-7b1ca6fbeb69)

And another: 

![Image](https://github.com/user-attachments/assets/e83509e0-14f0-426e-b8c7-25ba42bd84d2)

I get several when searching for IBM PCjr on Google Images, all showing brown.

**Comment by OBattler** [comment]
But, I can add the brown-less option, no problem, we already support that on the CGA.

**Comment by Jkapp76** [comment]
Every other port of this game has yellow soldiers. Even the composite CGA has the same colors with yellow. Commodore 64 as well.

The image on the back of the game box itself shows PCJr with yellow soldiers. 

There are definitely some games that need brown and some that need yellow. This color makes Roger in Space Quest's hair brown or blonde. And there's quite some info on Roger's dual hair colors.

**Comment by Jkapp76** [comment]
[CASWOLF.zip](https://github.com/user-attachments/files/20825994/CASWOLF.zip)

Here's the bootable disk image using PCDos2.1 (the EXE has been patched)

**Comment by Jkapp76** [comment]
[CASWOLF HD.zip](https://github.com/user-attachments/files/20826009/CASWOLF.HD.zip)
Here is the game patched to run from a hard drive.

**Comment by OBattler** [comment]
> The image on the back of the game box itself shows PCJr with yellow soldiers.

Do you have a link to that game box?

**Comment by eddmanx** [comment]
> [CASWOLF.zip](https://github.com/user-attachments/files/20825994/CASWOLF.zip)
> 
> Here's the bootable disk image using PCDos2.1 (the EXE has been patched)

I meant an untouched image from the original floppy. The patch shouldn't be needed when running on DOS 2.

**Comment by OBattler** [comment]
![Image](https://github.com/user-attachments/assets/b8f5c350-65e4-424d-8be4-ff28dbdc873b)

Interesting, I found this photo of a PCjr monitor with dark yellow instead of brown.

**Comment by Jkapp76** [comment]
I can send the un-patched disk soon when I'm back home.

And, I agree some games or maybe most games use the brown. But castle Wolfenstein and jumpman seem to be good examples of exceptions. And dosbox staging has documented this. And gave color options.

**Comment by Jkapp76** [comment]
[CASWOLF (orig).zip](https://github.com/user-attachments/files/20827259/CASWOLF.orig.zip)
Here's the original unpatched floppy version.

**Comment by Jkapp76** [comment]
John Novak from the DOSBox Staging github said the color depended on the monitor, not the computer... although PCJr did have two different color palettes depending on the age of the hardware. Certain monitors got a lighter brown that was more orange or dark yellow. He's the one who added the feature to Staging to switch between either option, as some games made use of either color. It's a simulated knob that varies from brown to a light yellow at the other extreme by choosing a number from 0 to some higher number. Somewhere in the middle is best for this alternate color, a dark yellow.

**Comment by Jkapp76** [comment]
![Image](https://github.com/user-attachments/assets/69ebbfc2-3571-4a6e-b587-6629375c2d7c)
Here's Page 20 from PCJr magazine itself, 1984 issue 10. Clearly shows dark yellow soldiers right in PCJr magazine. (archive.org)
There's another issue of this magazine that shows the bombs from jumpman this same yellow color instead of brown as well.

**Comment by eddmanx** [comment]
> [CASWOLF (orig).zip](https://github.com/user-attachments/files/20827259/CASWOLF.orig.zip) Here's the original unpatched floppy version.

This doesn't look to be original. Was the disk that you bought already bootable?

**Comment by Jkapp76** [comment]
Castle wolfenstein is a dos game that says right in the manual works best with dos 2.1. The game even included a patch right on the disk for 2.1 that would update something. But the game needed Dos added to the disk, which was done to this disk. (the update mentioned above is an unrelated update)

The sequel, "beyond castle wolfenstein" was a legit booter that did not use dos but booted it's own OS right from this disk.

**Comment by eddmanx** [comment]
Well it didn't exactly need it. Per the manual you could launch the game exe directly from the floppy without making it bootable.

Also, it seems the exe was already replaced with the cracked one by the previous user.

**Comment by Jkapp76** [comment]
I don't believe this CD.EXE is cracked in any way in this "orig" version I posted. It still requires the python patch to work correctly on PCJr from hard drive.

**Comment by eddmanx** [comment]
I'm not 100% sure, but it has a 2001 date and matches the other exes found on the internet. If you are not the one that replaced it, then maybe the floppy you got is modified. The old crack wasn't made properly anyway, so maybe it has issues running on PCjr, but not XT.

**Comment by Jkapp76** [comment]
I'm not completely sure about it... but, the old patch says it patches pcdos2.1 in some way and does not affect the game's exe. I believe the exe is the same as the one commonly found online regardless of that file date. 

I'm more fascinated with the colors issue. The game seems to boot. Seems to now run from hard drive after the python patch. I have screen shots of the end-game too, so it plays to the end. But that yellow\brown is interesting. The game does seem to run a touch too fast as well, the men talking sound like chipmunks a bit, but this is possible normal?

**Comment by eddmanx** [comment]
> I'm not completely sure about it... but, the old patch says it patches pcdos2.1 in some way and does not affect the game's exe. 

That DOS patch is only related to fixing a keyboard issue and is intended only for PCjr, per the readme. It patches a DOS file, and does nothing to the game exe. It's also unrelated to the cracked exes floating around the web.

~~Actually, I just tried the patch and patched a floppy of PC-DOS 2.1, booted from it, and now your orig image works on PCjr~~ without needing to apply that new python patch. So it seems the issue was an incompatibility between the game and DOS on PCjr.

EDIT: Well, turns out it has nothing to do with that patch either. Previously I was testing with MS-DOS 2.11 booting from the game floppy itself. I just booted from an unpatched PC-DOS 2.1 on PCjr and the game nonetheless works. I guess it's MS-DOS that is the culprit on PCjr, or maybe version 2.11 specifically.

> I believe the exe is the same as the one commonly found online regardless of that file date.

Yes, because basically everyone distributes the cracked exe. Some packages (e.g. the one from total dos collection) comes with two exe files, **CW.EX_** (seemingly the OG untouched file) and **CW.EXE**, and the latter matches the exe from your image, so if I had to guess your copy comes with the old cracked exe. Hopefully you didn't pay too much for it. I haven't yet seen an untouched image of this game anywhere.

**Comment by OBattler** [comment]
It's a common problem with old games that "original" copies floating around are in fact patched. Was the case even with something as common as Windows NT 3.1 RTM, of which, until fairly recently, the only copy floating around was the one with the patched Setup files to bypass the CPU detection for purposes of running it under virtualization, which however broke it on emulators (or on real old hardware).

**Comment by Jkapp76** [comment]
The main issue I raised seems to be solved. The joystick doesn't seem to work, but it won't work on dosbox either.

The info about the brown or yellow color was largely discussed on the staging github. From everything I know and saw discussed there I'd say dark yellow should be an option for the handful of games that were intended to use that color. Maybe where you choose the RGB or composite video modes we could add an additional RGB mode there?

**Comment by OBattler** [comment]
Yes, that's exactly what I was planning to, much the same it's already done for CGA.

**Comment by OBattler** [comment]
This is now done.

**Comment by Jkapp76** [comment]
I tested it. Looks great!
